File: leetcode/two-pointers/19.Remove-Nth-Node-From-End-of-List/19.Remove-Nth-Node-From-End-of-List.go
package leetcode

import (
	"github.com/yuancf1024/algorithm-go/structures"
)

type ListNode = structures.ListNode

/*
给你一个链表，删除链表的倒数第 n 个结点，并且返回链表的头结点。
1 <= n <= sz（链表节点数）=> 无需考虑 n 大于链表节点数的情形
*/


/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// 该实现有bug，没有把所有的边界条件考虑清楚。通过测试用例：190 / 208
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// 通过测试用例：190 / 208
func removeNthFromEnd_v1(head *ListNode, n int) *ListNode {
    if head == nil || head.Next == nil {
        return nil
    }

    slow, fast := head, head
    // for n > 0 { 
    //     fast = fast.Next
    //     n--
    // }

    // for fast.Next != nil { // 发生报错：panic: runtime error: invalid memory address or nil pointer dereference
    //     slow = slow.Next
    //     fast = fast.Next 
    // }

    for fast != nil { // 无法删除头节点，**因此需要新建一个头节点，其指针指向head**，参考解法1
        if n < 0 {
            slow = slow.Next
        }
        n--
        fast = fast.Next
    }
    slow.Next = slow.Next.Next
    return head
}

/*
这道题比较简单，先循环一次拿到链表的总长度，
然后循环到要删除的结点的前一个结点开始删除操作。
需要注意的一个特例是，有可能要删除头结点，要单独处理。

这道题有一种特别简单的解法。设置 2 个指针，
一个指针距离前一个指针 n 个距离。同时移动 2 个指针，
2 个指针都移动相同的距离。当一个指针移动到了终点，
那么前一个指针就是倒数第 n 个节点了。
*/

// 解法1
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummyHead := &ListNode{Next: head} // 包含了删除头节点的case
	preSlow, slow, fast := dummyHead, head, head
	for fast != nil {
		if n <= 0 {
			preSlow = slow
			slow = slow.Next
		}
		n--
		fast = fast.Next
	}
	preSlow.Next = slow.Next // 删除节点
	return dummyHead.Next
}

// 解法二， 对于n 大于链表节点数的特殊情形也可以通过
func removeNthFromEnd_v2(head *ListNode, n int) *ListNode {
	if head == nil {
		return nil
	}

	if n <= 0 {
		return head
	}

	current := head
	len := 0
	for current != nil {
		len++
		current = current.Next
	}
	if n > len {
		return head
	}

	if n == len {
		current := head
		head = head.Next
		current.Next = nil
		return head
	}

	current = head
	i := 0
	for current != nil {
		if i == len - n - 1 {
			deleteNode := current.Next
			current.Next = current.Next.Next
			deleteNode.Next = nil 
			break
		}
		i++
		current = current.Next
	}
	return head
}

File: leetcode-cpp/19.Remove-Nth-Node-From-End-of-List/19.Remove-Nth-Node-From-End-of-List.cpp
/* 19. 删除链表的倒数第 N 个结点

给你一个链表，删除链表的倒数第 n 个结点，并且返回链表的头结点。

示例 1：
输入：head = [1,2,3,4,5], n = 2
输出：[1,2,3,5]

示例 2：
输入：head = [1], n = 1
输出：[]

示例 3：
输入：head = [1,2], n = 1
输出：[1]
 
提示：

链表中结点的数目为 sz
1 <= sz <= 30
0 <= Node.val <= 100
1 <= n <= sz
 
进阶：你能尝试使用一趟扫描实现吗？

*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// 双指针的经典应用 参考代码随想录
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummyHead = new ListNode(0);
        dummyHead->next = head;
        ListNode* slow = dummyHead;
        ListNode* fast = dummyHead;
        while (n-- && fast != NULL) { // fast先走n步
            fast = fast->next;
        }
        fast = fast->next; // fast再提前走一步，因为需要让slow指向删除节点的上一个节点
        while (fast != NULL) {
            fast = fast->next;
            slow = slow->next;
        }
        slow->next = slow->next->next;
        return dummyHead->next;
    }
};

/*
思路
双指针的经典应用，如果要删除倒数第n个节点，让fast移动n步，
然后让fast和slow同时移动，直到fast指向链表末尾。
删掉slow所指向的节点就可以了。

思路是这样的，但要注意一些细节。

分为如下几步：

- 首先这里我推荐大家使用虚拟头结点，这样方便处理删除实际头结点的逻辑，

- 定义fast指针和slow指针，初始值为虚拟头结点，如图：

- fast首先走n + 1步 ，为什么是n+1呢，因为只有这样同时移动的
时候slow才能指向删除节点的上一个节点（方便做删除操作），如图：

- fast和slow同时移动，直到fast指向末尾，如题：

- 删除slow指向的下一个节点，如图：
*/

